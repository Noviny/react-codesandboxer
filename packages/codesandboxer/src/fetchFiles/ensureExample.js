// @flow

import fetchRelativeFile from '../fetchRelativeFile';
import replaceImports from '../replaceImports';
import absolutesToRelative from '../utils/absolutesToRelative';
import parseFile from '../parseFile';

import type { Import, Package, GitInfo, Config } from '../types';

export default async function ensureExample(
  example?: string | Promise<string>,
  importReplacements: Array<Import>,
  pkg: Package,
  examplePath: string,
  gitInfo: GitInfo,
  config: Config,
) {
  if (example) {
    let exampleContent = await Promise.resolve(example);
    let content = replaceImports(
      exampleContent,
      importReplacements.map(m => [
        absolutesToRelative(examplePath, m[0]),
        m[1],
      ]),
    );
    return parseFile(content, pkg);
  } else {
    return fetchRelativeFile(
      examplePath,
      pkg,
      importReplacements,
      gitInfo,
      config,
    );
  }
}
