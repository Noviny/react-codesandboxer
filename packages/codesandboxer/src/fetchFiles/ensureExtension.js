// @flow
import path from 'path-browserify';

export default function ensureExtension(filePath: string): string {
  if (!path.extname(filePath)) {
    return `${filePath}.js`;
  } else {
    return filePath;
  }
}
