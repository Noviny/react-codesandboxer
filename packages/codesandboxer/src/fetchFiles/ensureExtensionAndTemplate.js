// @flow
const ensureExtensionAndTemplate = (
  extension: string,
  extensions: string[] = [],
  template?: 'create-react-app' | 'create-react-app-typescript' | 'vue-cli'
): {
  extensions: string[],
  template: 'create-react-app' | 'create-react-app-typescript' | 'vue-cli',
} => {
  let extensionsSet = new Set(['.js', '.json', ...extensions]);
  extensionsSet.add(extension);

  if (
    ['.ts', '.tsx'].includes(extension) ||
    template === 'create-react-app-typescript'
  ) {
    if (!template) template = 'create-react-app-typescript';
    extensionsSet.add('.ts');
    extensionsSet.add('.tsx');
  }

  if (extension === '.vue' || template === 'vue-cli') {
    if (!template) template = 'vue-cli';
    extensionsSet.add('.vue');
  }

  if (!template) {
    template = 'create-react-app';
  }

  return { extensions: [...extensionsSet], template };
};

export default ensureExtensionAndTemplate;

/*
That other implementation:

  if (!config.template) {
    if (['.ts', '.tsx'].includes(extension)) {
      config.template = 'create-react-app-typescript';
    } else if (extension === '.vue' && !config.template) {
      config.template = 'vue-cli';
    } else {
      config.template = 'create-react-app';
    }
  }

  let extensions = ['.js', '.json'];
  if (config.extensions) extensions = [...extensions, ...config.extensions];
  if (
    extension &&
    !baseExtensions.includes(extension) &&
    !extensions.includes(extension)
  ) {
    extensions.push(extension);
  }
*/
