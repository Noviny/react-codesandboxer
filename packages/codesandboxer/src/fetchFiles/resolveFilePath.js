// @flow
export default function resolveFilePath(path: string): string {
  let fileMatch = path.match(/.+(\..+)$/);
  if (!fileMatch) {
    return `${path}.js`;
  } else {
    return path;
  }
}
