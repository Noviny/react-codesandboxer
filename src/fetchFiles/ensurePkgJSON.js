// @flow
import fetchRelativeFile from '../fetchRelativeFile';
import type { Package, Import, FetchConfig } from '../types';

export default async function ensurePKGJSON(
  maybePkg?: Package | string | Promise<Package | string>,
  importReplacements: Array<Import>,
  gitInfo: FetchConfig,
): Promise<Package> {
  let pkg = await Promise.resolve(maybePkg);
  if (typeof pkg === 'object') {
    return pkg;
  } else if (typeof pkg === 'string') {
    return fetchRelativeFile(
      pkg,
      // $FlowFixMe - we know here that this will not be a js file, the only time we NEED a pkg
      {},
      importReplacements,
      gitInfo,
    ).then(({ file }) => JSON.parse(file));
  } else if (!pkg) {
    return fetchRelativeFile(
      'package.json',
      // $FlowFixMe - we know here that this will not be a js file, the only time we NEED a pkg
      {},
      importReplacements,
      gitInfo,
    ).then(({ file }) => JSON.parse(file));
  } else throw new Error('could not understand passed in package.json');
}
