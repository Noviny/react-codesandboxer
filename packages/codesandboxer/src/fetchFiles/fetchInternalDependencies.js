// @flow
import type {
  Files,
  Package,
  GitInfo,
  ImportReplacement,
  Config,
} from '../types';

import fetchRelativeFile from '../fetchRelativeFile';
import ensureExtension from './ensureExtension';
import resolvePath from '../utils/resolvePath';

export default async function fetchInternalDependencies(
  internalImports: Array<string>,
  files: Files,
  pkg: Package,
  deps: { [string]: string },
  gitInfo: GitInfo,
  importReplacements: Array<ImportReplacement>,
  config: Config,
  accumulatedInternalDependencies: string[] = [],
) {
  let newFiles = await Promise.all(
    internalImports.map(path =>
      fetchRelativeFile(path, pkg, importReplacements, gitInfo, config),
    ),
  );

  let moreInternalImports = [];
  for (let f of newFiles) {
    files[ensureExtension(f.path)] = { content: f.file };
    deps = { ...deps, ...f.deps };
    f.internalImports.forEach(m =>
      moreInternalImports.push(resolvePath(f.path, m)),
    );
  }

  accumulatedInternalDependencies = accumulatedInternalDependencies.concat(
    internalImports,
  );

  moreInternalImports = moreInternalImports.filter(
    mpt => !accumulatedInternalDependencies.includes(mpt),
  );

  if (moreInternalImports.length > 0) {
    let moreFiles = await fetchInternalDependencies(
      moreInternalImports,
      files,
      pkg,
      deps,
      gitInfo,
      importReplacements,
      config,
      accumulatedInternalDependencies,
    );
    return {
      files: { ...files, ...moreFiles.files },
      deps: { ...deps, ...moreFiles.deps },
    };
  } else {
    return { files, deps };
  }
}
