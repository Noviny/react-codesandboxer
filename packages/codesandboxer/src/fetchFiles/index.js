// @flow
import resolvePath from '../utils/resolvePath';
import replaceImports from '../replaceImports';

import ensureExample from './ensureExample';
import ensurePKGJSON from './ensurePkgJSON';
import fetchInternalDependencies from './fetchInternalDependencies';

import type { Package, GitInfo, Dependencies, Import, Files } from '../types';
import { baseFiles } from '../constants';

export default async function({
  examplePath,
  pkgJSON,
  gitInfo,
  importReplacements,
  example,
  allowJSX,
}: {
  examplePath: string,
  pkgJSON?: Package | string | Promise<Package | string>,
  gitInfo: GitInfo,
  importReplacements: Array<Import>,
  dependencies?: Dependencies,
  providedFiles?: Files,
  example?: string | Promise<string>,
  name?: string,
  allowJSX?: boolean,
}) {
  let config = { allowJSX: !!allowJSX };
  let pkg = await ensurePKGJSON(pkgJSON, importReplacements, gitInfo, config);

  let { file, deps, internalImports } = await ensureExample(
    example,
    importReplacements,
    pkg,
    examplePath,
    gitInfo,
    config,
  );

  let files = {
    ...baseFiles,
    'example.js': {
      content: replaceImports(
        file,
        internalImports.map(m => [m, `./${resolvePath(examplePath, m)}`]),
      ),
    },
  };

  let final = await fetchInternalDependencies(
    internalImports.map(m => resolvePath(examplePath, m)),
    files,
    pkg,
    deps,
    gitInfo,
    importReplacements,
    config,
  );
  return final;
}
