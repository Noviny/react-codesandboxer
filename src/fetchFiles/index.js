// @flow
import resolvePath from '../utils/resolvePath';
import replaceImports from '../replaceImports';

import resolveFilePath from './resolveFilePath';
import ensureExample from './ensureExample';
import ensurePKGJSON from './ensurePkgJSON';
import fetchInternalDependencies from './fetchInternalDependencies';

import type {
  Package,
  FetchConfig,
  Dependencies,
  Import,
  Files,
} from '../types';
import finaliseCSB from './finaliseCSB';
import { baseFiles } from '../constants';

export default async function({
  examplePath,
  pkgJSON,
  gitInfo,
  importReplacements,
  dependencies = {},
  providedFiles = {},
  example,
}: {
  examplePath: string,
  pkgJSON?: Package | string | Promise<Package | string>,
  gitInfo: FetchConfig,
  importReplacements: Array<Import>,
  dependencies?: Dependencies,
  providedFiles?: Files,
  example?: string | Promise<string>,
}) {
  let pkg = await ensurePKGJSON(pkgJSON, importReplacements, gitInfo);

  let { file, deps, internalImports } = await ensureExample(
    example,
    importReplacements,
    pkg,
    examplePath,
    gitInfo,
  );

  let files = {
    ...baseFiles,
    'example.js': {
      content: replaceImports(
        file,
        internalImports.map(m => [m, `./${resolvePath(examplePath, m)}`]),
      ),
    },
  };

  let final = await fetchInternalDependencies(
    internalImports.map(m => resolvePath(examplePath, m)),
    files,
    pkg,
    deps,
    gitInfo,
    importReplacements,
  );
  return finaliseCSB(final, providedFiles, dependencies);
}
