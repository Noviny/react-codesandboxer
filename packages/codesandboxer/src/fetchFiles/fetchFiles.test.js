import fetchFiles from './index';
/* eslint-disable-next-line no-unused-vars */
import isomorphic from 'isomorphic-unfetch';

const BBInfo = {
  account: 'atlassian',
  repository: 'atlaskit-mk-2',
  branch: '6546190ec6d8e1e47566882177fa941bcb8bf576',
  host: 'bitbucket',
};

const getMainObj = (userOpts = {}) => ({
  examplePath: 'packages/elements/avatar/examples/01-basicAvatar.js',
  pkgJSON: 'packages/elements/avatar/package.json',
  gitInfo: BBInfo,
  importReplacements: [['packages/elements/avatar/src', '@atlaskit/avatar']],
  dependencies: { '@atlaskit/avatar': 'latest' },
  providedFiles: {},
  ...userOpts,
});

const getSandboxerObj = (userOpts = {}) => ({
  examplePath: 'fixtures/simple.js',
  pkgJSON: 'package.json',
  gitInfo: {
    account: 'Noviny',
    repository: 'codesandboxer',
    branch: 'master',
    host: 'github',
  },
  dependencies: { '@atlaskit/avatar': 'latest' },
  providedFiles: {},
  ...userOpts,
});

it('should fetch an example from atlaskit', () => {
  return fetchFiles(getMainObj()).then(res => {
    expect(res.files).toMatchSnapshot();
  });
});
it('should fetch our basic fixture example', () => {
  return fetchFiles(getSandboxerObj()).then(res => {
    expect(res.files).toMatchSnapshot();
  });
});
it('should fetch a css example from our fixtures', () => {
  return fetchFiles(
    getSandboxerObj({ examplePath: 'fixtures/withCssImport.js' }),
  ).then(res => {
    expect(res.files).toMatchSnapshot();
  });
});
