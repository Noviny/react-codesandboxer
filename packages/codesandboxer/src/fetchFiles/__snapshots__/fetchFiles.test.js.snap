// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should fetch an example from atlaskit 1`] = `
Object {
  "example.js": Object {
    "content": "// @flow
import React from 'react';
import Avatar from '@atlaskit/avatar';
import { Block, Gap } from './packages/elements/avatar/examples-util/helpers';

export default () => (
  <div>
    <Block heading=\\"Circle\\">
      <Avatar name=\\"xxlarge\\" size=\\"xxlarge\\" />
      <Gap />
      <Avatar name=\\"xlarge\\" size=\\"xlarge\\" presence=\\"online\\" />
      <Gap />
      <Avatar name=\\"large\\" size=\\"large\\" presence=\\"offline\\" />
      <Gap />
      <Avatar name=\\"medium\\" size=\\"medium\\" presence=\\"busy\\" />
      <Gap />
      <Avatar name=\\"small\\" size=\\"small\\" presence=\\"focus\\" />
      <Gap />
      <Avatar name=\\"xsmall\\" size=\\"xsmall\\" />
    </Block>
    <Block heading=\\"Square\\">
      <Avatar appearance=\\"square\\" name=\\"xxlarge\\" size=\\"xxlarge\\" />
      <Gap />
      <Avatar
        appearance=\\"square\\"
        name=\\"xlarge\\"
        size=\\"xlarge\\"
        status=\\"approved\\"
      />
      <Gap />
      <Avatar appearance=\\"square\\" name=\\"large\\" size=\\"large\\" status=\\"declined\\" />
      <Gap />
      <Avatar appearance=\\"square\\" name=\\"medium\\" size=\\"medium\\" status=\\"locked\\" />
      <Gap />
      <Avatar appearance=\\"square\\" name=\\"small\\" size=\\"small\\" />
      <Gap />
      <Avatar appearance=\\"square\\" name=\\"xsmall\\" size=\\"xsmall\\" />
    </Block>
  </div>
);
",
  },
  "index.html": Object {
    "content": "<div id=\\"root\\"></div>",
  },
  "index.js": Object {
    "content": "/**
  This CodeSandbox has been automatically generated using
  \`react-codesandboxer\`. If you're curious how that happened, you can
  check out our docs here: https://github.com/noviny/react-codesandboxer

  If you experience any struggles with this sandbox, please raise an issue
  on github. :)
*/
import React from 'react';
import ReactDOM from 'react-dom';
import Example from './example';

ReactDOM.render(
<Example />,
document.getElementById('root')
);",
  },
  "package.json": Object {
    "content": "{
  \\"name\\": \\"react-codesandboxer-example\\",
  \\"version\\": \\"0.0.0\\",
  \\"description\\": \\"A simple example deployed using react-codesandboxer\\",
  \\"main\\": \\"index.js\\",
  \\"dependencies\\": {
    \\"react\\": \\"15.6.0 - 16.2.0\\",
    \\"styled-components\\": \\"1.4.6 - 3\\",
    \\"@atlaskit/theme\\": \\"^2.4.0\\",
    \\"@atlaskit/avatar\\": \\"latest\\",
    \\"react-dom\\": \\"15.6.0 - 16.2.0\\"
  }
}",
  },
  "packages/elements/avatar/examples-util/helpers.js": Object {
    "content": "// @flow
import React from 'react';
import type { Node } from 'react';
import styled from 'styled-components';
import { colors, math, gridSize } from '@atlaskit/theme';

const Wrapper = styled.div\`
  margin-top: \${gridSize}px;
\`;

const ChildrenWrapper = styled.div\`
  align-items: baseline;
  color: \${colors.text};
  display: flex;

  > * {
    margin-right: \${gridSize}px;
  }
\`;

export const Note = styled.p\`
  color: \${colors.N100};
  font-size: \${props => (props.size === 'large' ? '1.15em' : '0.9rem')};
  margin-top: \${math.divide(gridSize, 2)}px;
  margin-bottom: \${math.multiply(gridSize, 2)}px;
\`;

export const Code = styled.code\`
  background-color: \${colors.R50};
  border-radius: 0.2em;
  color: \${colors.R400};
  font-size: 0.85em;
  line-height: 1.1;
  padding: 0.1em 0.4em;
\`;

export const Gap = styled.span\`
  margin-right: \${gridSize}px;
\`;
// $FlowFixMe
export const Dot = styled(Gap)\`
  height: \${math.multiply(gridSize, 3)}px;
  width: \${math.multiply(gridSize, 3)}px;
\`;
export const Heading = styled.div\`
  color: \${colors.subtleHeading};
  display: flex;
  font-size: 0.8rem;
  font-weight: 500;
  margin-bottom: 0.5em;
  text-transform: uppercase;
\`;

export const Block = ({
  children,
  heading,
}: {
  children: ?Node,
  heading?: string,
}) => (
  <Wrapper>
    {heading ? <Heading>{heading}</Heading> : null}
    <ChildrenWrapper>{children}</ChildrenWrapper>
  </Wrapper>
);
",
  },
}
`;

exports[`should fetch an example from react-select 1`] = `
Object {
  "example.js": Object {
    "content": "// @flow
import React from 'react';
import Avatar from '@atlaskit/avatar';
import { Block, Gap } from './packages/elements/avatar/examples-util/helpers';

export default () => (
  <div>
    <Block heading=\\"Circle\\">
      <Avatar name=\\"xxlarge\\" size=\\"xxlarge\\" />
      <Gap />
      <Avatar name=\\"xlarge\\" size=\\"xlarge\\" presence=\\"online\\" />
      <Gap />
      <Avatar name=\\"large\\" size=\\"large\\" presence=\\"offline\\" />
      <Gap />
      <Avatar name=\\"medium\\" size=\\"medium\\" presence=\\"busy\\" />
      <Gap />
      <Avatar name=\\"small\\" size=\\"small\\" presence=\\"focus\\" />
      <Gap />
      <Avatar name=\\"xsmall\\" size=\\"xsmall\\" />
    </Block>
    <Block heading=\\"Square\\">
      <Avatar appearance=\\"square\\" name=\\"xxlarge\\" size=\\"xxlarge\\" />
      <Gap />
      <Avatar
        appearance=\\"square\\"
        name=\\"xlarge\\"
        size=\\"xlarge\\"
        status=\\"approved\\"
      />
      <Gap />
      <Avatar appearance=\\"square\\" name=\\"large\\" size=\\"large\\" status=\\"declined\\" />
      <Gap />
      <Avatar appearance=\\"square\\" name=\\"medium\\" size=\\"medium\\" status=\\"locked\\" />
      <Gap />
      <Avatar appearance=\\"square\\" name=\\"small\\" size=\\"small\\" />
      <Gap />
      <Avatar appearance=\\"square\\" name=\\"xsmall\\" size=\\"xsmall\\" />
    </Block>
  </div>
);
",
  },
  "index.html": Object {
    "content": "<div id=\\"root\\"></div>",
  },
  "index.js": Object {
    "content": "/**
  This CodeSandbox has been automatically generated using
  \`react-codesandboxer\`. If you're curious how that happened, you can
  check out our docs here: https://github.com/noviny/react-codesandboxer

  If you experience any struggles with this sandbox, please raise an issue
  on github. :)
*/
import React from 'react';
import ReactDOM from 'react-dom';
import Example from './example';

ReactDOM.render(
<Example />,
document.getElementById('root')
);",
  },
  "package.json": Object {
    "content": "{
  \\"name\\": \\"react-codesandboxer-example\\",
  \\"version\\": \\"0.0.0\\",
  \\"description\\": \\"A simple example deployed using react-codesandboxer\\",
  \\"main\\": \\"index.js\\",
  \\"dependencies\\": {
    \\"react\\": \\"15.6.0 - 16.2.0\\",
    \\"styled-components\\": \\"1.4.6 - 3\\",
    \\"@atlaskit/theme\\": \\"^2.4.0\\",
    \\"@atlaskit/avatar\\": \\"latest\\",
    \\"react-dom\\": \\"15.6.0 - 16.2.0\\"
  }
}",
  },
  "packages/elements/avatar/examples-util/helpers.js": Object {
    "content": "// @flow
import React from 'react';
import type { Node } from 'react';
import styled from 'styled-components';
import { colors, math, gridSize } from '@atlaskit/theme';

const Wrapper = styled.div\`
  margin-top: \${gridSize}px;
\`;

const ChildrenWrapper = styled.div\`
  align-items: baseline;
  color: \${colors.text};
  display: flex;

  > * {
    margin-right: \${gridSize}px;
  }
\`;

export const Note = styled.p\`
  color: \${colors.N100};
  font-size: \${props => (props.size === 'large' ? '1.15em' : '0.9rem')};
  margin-top: \${math.divide(gridSize, 2)}px;
  margin-bottom: \${math.multiply(gridSize, 2)}px;
\`;

export const Code = styled.code\`
  background-color: \${colors.R50};
  border-radius: 0.2em;
  color: \${colors.R400};
  font-size: 0.85em;
  line-height: 1.1;
  padding: 0.1em 0.4em;
\`;

export const Gap = styled.span\`
  margin-right: \${gridSize}px;
\`;
// $FlowFixMe
export const Dot = styled(Gap)\`
  height: \${math.multiply(gridSize, 3)}px;
  width: \${math.multiply(gridSize, 3)}px;
\`;
export const Heading = styled.div\`
  color: \${colors.subtleHeading};
  display: flex;
  font-size: 0.8rem;
  font-weight: 500;
  margin-bottom: 0.5em;
  text-transform: uppercase;
\`;

export const Block = ({
  children,
  heading,
}: {
  children: ?Node,
  heading?: string,
}) => (
  <Wrapper>
    {heading ? <Heading>{heading}</Heading> : null}
    <ChildrenWrapper>{children}</ChildrenWrapper>
  </Wrapper>
);
",
  },
}
`;
